text/0000-roadmap-2018.md

```diff
@@ -0,0 +1,432 @@
+
+#### Language
+
+
+> [Rust 2018: Consolidation](https://deterministic.space/rust-2018.html) (@killercup)
+
+The most prominent language work in the pipeline stems from [2017’s ergonomics initiative](https://blog.rust-lang.org/2017/03/02/lang-ergonomics.html). Almost all of the accepted RFCs from the initiative are available on nightly, but polish, testing, and consensus work will take time:
+
+> [I’d like to reach a final decision to ship or drop all of the ergonomics RFCs that were accepted. I hope to see this completed over the next several months . . . I hope we can have a clearer (and more spaced out) schedule for this so that their FCPs are staggered.](https://boats.gitlab.io/blog/post/2018-01-08-goals-for-rust/) (@withoutboats)
+
+Among these productivity features are a few “headliners” that will form the backbone of the release:
+
+
+- **Non-lexical lifetimes**. Currently in “alpha” state on nightly, with work ongoing.
+- **`impl Trait`**. Nearing readiness for stabilization FCP.
+- **Generators**. “Beta” state on nightly, but some design issues need resolution.
+- **Module system changes**. Largely usable on nightly; will need testing, feedback, and bikeshedding.
+
+In addition, there are some other headlining features which are nearing stabilization and should ship prior to the epoch:
+
+
+- **SIMD**. The core SIMD intrinsics are nearing readiness for stabilization, and with luck we may be able to stabilize some vendor-agnostic primitives as well.
+- **Custom allocators**. The machinery has been in place for some time. Let's settle the details and ship.
+- **Macros 2.0**. This feature is implemented and working, but stabilization will require us to reach a comfort level with the handling of hygiene and a few other core issues.
+
+Between generators and macros 2.0, we will have *some* support for async/await on stable Rust (possibly using macros, possibly some other way).
+
+Finally, there are several highly-awaited features that are **unlikely to ship** **in the Rust 2018** **epoch release** **(though they may ship later in the year):**
+
+
+- **Generic associated types**. This feature will almost certainly land in *nightly* in 2018, and may even stabilize during the year. However, enough implementation work remains that it’s unlikely to be stable prior to the epoch release.
+- **Specialization**. Stabilization is blocked on a number of *extremely* subtle issues, including a revamp of the trait system and [finding a route to soundness](http://aturon.github.io/blog/2017/07/08/lifetime-dispatch/). We cannot afford to spend time on these issues until after the epoch release ships.
+- **const generics**. This feature is likely to land in *nightly* in 2018, but will not be ready to stabilize this year given the substantial work that remains.
```

Comment:
I personally don't think const generics should be deprioritized over custom allocators and SIMD, both of which also still require significant implementation effort.

Reply:
The scale of implementation work needed is probably two orders of magnitude different, though. This roadmap is focused on what the Rust team feels confident we can *ship* on the stated timeline.